### ISSUE TITLE
Garden Operator CI

### ISSUE THREAD

**Issue by gardener-robot-ci-1** [gardener/gardener#12]
**Issue by [kubernetes-jenkins](https://github.com/kubernetes-jenkins)**
_Wednesday Jul 19, 2017 at 18:44 GMT_
_Originally opened as https://git.removed/kubernetes-attic/garden-operator/issues/24_

----

# Stories
* As contributor I depend on a continuous integration/delivery pipeline, so that I neither waste my time with repetitive tasks nor can I make (fatal) mistakes.
 
# Motivation
Newly hired or joined colleagues or contributors from outside the group of colleagues who primarily work on a repo/subject/topic have a hard time contributing, especially when sitting in remote locations, but in order to sustainably grow, we need everybody to be happy and help our cause.

# Acceptance Criteria
- [ ] Continuous integration (acting on PRs/Merges) including the following steps (in chronological order, not the order of importance):
  - [ ] Optional (can be done in phase 4): Formal checks like referenced issue
  - [x] Optional (can be done in phase 3): Lint checks
  - [x] Mandatory (must be done in phase 1): Build
  - [x] Optional (can be done in phase 3): Unit tests
  - [ ] Optional (can be done in phase 3): Acceptance tests
  - [ ] Optional (can be done in phase 2): Deployment to staging
  - [ ] Optional (can be done in phase 2): Integration tests on staging
  - [ ] Mandatory (must be done in phase 1): Deployment to production
  - [ ] Mandatory (already done in phase 0): Integration tests on production

### Details (not complete, to be extended by all of us as we go along)
- [x] Deployment of Jenkins into new cluster
  - [x] Adapt Jenkins to new cluster: https://git.removed/kubernetes/jenkins-setup
    - New cluster: https://git.removed/kubernetes/cluster-infra-ci (https://api.ci.infra.k8s.sapcloud.io)
  - [x] Deployment of Helm chart with new values from https://git.removed/kubernetes/jenkins-setup
  - [x] Switch of Jenkins domain http://jenkins.k8s.sapcloud.io:30000/
- [x] Create secrets with kubernetes-jenkins password in new cluster (ideally automate that with e.g. jenkins-setup repo/script or the Jenkins-tools job/script), mount it into the pod/container template and adapt JenkinsFile accordingly
- [x] Deploy into Garden DEV cluster into separate NS (watching only this NS) and clone and run the garden-it against this NS
- [x] Run Jenkins pipeline also on PRs
- [ ] Update PR during pipeline execution on certain stages, e.g. Lint tests passed, Unit tests passed, Integration tests passed, etc.
- [ ] Exact plan of actual pipeline and how a change shall be prepared, what happens to the pull request, where it is tested, how the build result is promoted and deployed to production
- [ ] In-depth evaluation of similar tools like Concourse or Drone which are more promising than Jenkins + Kubernetes-Plugin

# Definition of Done
- [ ] Knowledge is distributed: Have you spread your knowledge in pair programming/code review?

:information_source: Migrated from Jira issue [KUBE-149](https://jtrack.removed/browse/KUBE-149)

**Comment by gardener-robot-ci-1** [comment]
**Comment by [vlerenc](https://github.com/vlerenc)**
_Sunday Jul 23, 2017 at 04:13 GMT_

----

Note by @d023280:

We have our very first test case (see commit ff0159b38dc0bead01bb2f7d584927bc04557af1) in the garden-operator and should call it in our integration server/pipeline with `make test`.

**Comment by gardener-robot-ci-1** [comment]
**Comment by [vlerenc](https://github.com/vlerenc)**
_Wednesday Aug 02, 2017 at 06:12 GMT_

----

Our pipeline should include the following:

On every dev codeline PR/update:
* Formal checks like referenced issue
* Lint checks
* Unit tests (mocking everything outside the unit/module)
* Acceptance tests (mocking everything outside the component/GitHub project)

On every release codeline PR/update:
* Deployment into a Garden DEV cluster for testing purposes, not watching all namespaces, but only an integration-test namespace:
  * Create, consume, and delete a cluster on all infrastructure in parallel
  * Take an existing cluster and update it

Testing matrix:
* One dimension is the IaaS, i.e. AWS, Azure, GCP, and OpenStack
* Another dimension is creation respectively update (from the latest operator version for that particular Kubernetes version) of the various supported versions (maybe we can limit creation to fewer versions than update)

| | AWS | Azure | GCP | OpenStack |
|---|---|---|---|---|
Creation of new v1.9 | ☀️ | ❌ | ☀️ | ❌ |
Creation of new v1.8 | ☀️ | ☀️ | ☀️ | ☀️ |
Update to latest v1.9 | ☀️ | ☀️ | ☀️ | ❌ |
Update to latest v1.8 | ☀️ | ❌ | ☀️ | ❌ |
Update to latest v1.7 | ❌ | ☀️ | ☀️ | ☀️ |
Update to latest v1.6 | ☀️ | ☀️ | ☀️ | ❌ |

After it gets deployed to production:
* Start first with Canary updates (single clusters before going crazy on all clusters); if we have designated namespaces, start with the less critical ones, maybe internal/our, but not the seed clusters, of course, which should come last
* Scale out to all namespaces and do that in parallel, but still limited (no more than a configurable percentage, e.g. 20%)

**Comment by gardener-robot-ci-1** [comment]
**Comment by [vlerenc](https://github.com/vlerenc)**
_Tuesday Sep 05, 2017 at 14:58 GMT_

----

Since Jenkins, more so Jenkins pipelines, and especially the Jenkins Kubernetes Plugin comes with significant limitations and actually breaks the main argument for Jenkins with its ton of plugins and the huge community (only little of that is available in the execution pods), **Dirk will split off and try to implement the pipeline with Drone** which he will deploy to the same cluster. **Martin would be interested to do the same with Prow in the same cluster**. Then we could compare the solutions. The main work we do now is anyhow in the concept and design of the pipeline and in the images and scripts. The expectation is, that we can reuse this work even if we decide to switch from Jenkins to Drone, Prow or another solution.

**Comment by gardener-robot-ci-1** [comment]
**Comment by [vlerenc](https://github.com/vlerenc)**
_Tuesday Sep 05, 2017 at 15:30 GMT_

----

From our workshop in early August:
![20170801_181334](https://git.removed/storage/user/4487/files/b1089e40-925f-11e7-95f2-e1b1448e3daf)
  - Branching
    - Stay with 3 major branches:
      - Devlop->Release->Master          
      - see 20170801_181334.jpg
    - Team members from corresponding component can push to Devlop directly
      (or fork and open pull request if they like to (e.g. for major changes/refactorings))
    - Contributors (colleagues who work most of their time on a different component)    
      have to fork and create a pull request
    - For release prepartion and qualification release branch is created/updated
      to corresponding community
      - Qualification takes place. Either in staging landscape (if exists) or using
      a Canary appraoch with a phased rollout on productive landscape
    - Upon positive release decision master branch is set to Release branch (so that
      master always reflect current released/productive running software)
    - Hotfixing undefined yet.
     - Options are:
      - Create a HotFix branch based on master to create and qualify HotFix
      - Re-Use Release branch (in case it's not taken for upcoming release qualification yet)
        - If so, skiping/stop upcoming release or (in case qualification is almost done)
        add HotFix to next (planned) release and deploy it with next release
  - Build for Develop branch creates release ready Artifact (Docker Image) which
    is used for any further qualification ("Build once")
     - Build in Develop branch is triggered on any pushed commit on Develop
     - Additional Builds for all pull requests (targeting Develop) shall be conducted
  - Add more and more testing, the further the code moves in the deployment pipeline
    (->branches)
    - see 20170801_170115.jpg
    - see https://git.removed/kubernetes/garden-operator/issues/24#issuecomment-764226
  - Versioning:
    - See desicion on versioning https://git.removed/kubernetes/garden-operator/issues/56

**Comment by gardener-robot-ci-1** [comment]
**Comment by [d036774](https://github.com/d036774)**
_Tuesday Sep 05, 2017 at 15:47 GMT_

----

@d043832 @d051236 @d065849 
I just received this:
„….. some limitation with the usage of the kubernetes-plugin (up to version 0.12)……For now it is only possible to have 5 concurrent connections to Kubernetes cluster”
from the [Jenkins as a Service JAM Group](https://jam.removed/c/jamatsap.com/feed/item/rOAZzbO6rJEcV98fEjM5V8)

I cannot reach our Jenkins right now, but I think we have to ensure that we have at least the 0.12 version of the K8s plugin, as it contains this fix:
https://github.com/jenkinsci/kubernetes-plugin/commit/3997aff4555b36c65d2c93a500958bf3db945b5b

**Comment by gardener-robot-ci-1** [comment]
**Comment by [d036774](https://github.com/d036774)**
_Monday Sep 11, 2017 at 14:22 GMT_

----

Results from [Drone](https://drone.io/) evaluation:     

- Evaluation stop for now (11.09.2017)
   - Mail from Vedran 11.09.:
``` 
....
Dirk also looked into Drone and noticed other problems, hence no viable option either (at least for now):

    Not integrated with Kubernetes; containers are launched through Docker directly from an agent that can only run one build at a time (worse than Jenkins and certainly Prow)
    Version 0.x that is still incompatibly changed from one minor version to another (worse than Jenkins)
    Poor documentation (same as Jenkins)
    Popular/Growth
    Nice/modern UI (better than Jenkins and Prow)

Not enough reasons to go for Drone right now.
....
```
Some more details:
- Drone consists of one main server (docker image) + multiple agents (docker images)
  - Main server runs the UI and integrates e.g. with GitHub
  - Upon receiving events (e.g. from GitHub or the CLI client) it schedules a new build on the next free agent.
  - Agent clones repository and executes given build steps
  - Each build step is executed in an own Docker container which is a fork/a sibling of the agent container
- [Drone Issue #1815 - Discussion among Drone developers/users to enable Kubernetes as runtime engine instead of docker only.](https://github.com/drone/drone/issues/1815)
- [NY Times Success story on using Drone](https://open.blogs.nytimes.com/2017/01/12/continuous-deployment-to-google-cloud-platform-with-drone/)
- (Out-of-date) [Examples how to run Drone on Kubernetes](https://github.com/drone-demos/drone-on-kubernetes) 
- [Gogs and Drone on Kuberntes](http://larmog.github.io/2016/02/07/gogs-and-drone-on-kubernetes-arm---part-1/)

**Comment by vlerenc** [comment]
Done internally by other subteam for now until we find the time to expose it properly.
